### ISSUE TITLE
Grafeas integration

### ISSUE THREAD

**Issue by furuholm** [quay/clair#493]
We are looking at [Grafeas](https://grafeas.io) for managing our software supply chain metadata. As vulnerability scanning is a key component in this process I would like to integrate Clair with Grafeas. I'm currently working at doing such an integration by writing a third party client but I was thinking that maybe this could be done better by doing the integration directly between Clair and Grafeas instead.

I don't have a proposal for how to best do this (or if it is even suitable), but if there are others that are interested maybe this could be the start of a discussion.

**Comment by jzelinskie** [comment]
I've actually spoken and worked with the Grafeas team and I think an integration between Grafeas and Clair is a great idea.

I need to take a deep dive into their codebase, because it might make sense that Clair not only tag Artifact metadata in Grafeas, but also use it as a backing store rather than something like Postgres.

@R2wenD2 A penny for your thoughts?

**Comment by R2wenD2** [comment]
Yes! Grafeas is a suitable backing store for Clair. I'd be happy to collaborate here.

**Comment by furuholm** [comment]
Excellent!

Regarding Grafeas as a backend for Clair I think that Grafeas would be a great backend for CVE:s (both notes and occurrences in Grafeas lingo). But Clair stores other things as well, like data related to notifications and more. I don't think this is something that Grafeas is suitable for as Grafeas is not a general purpose database. So while I am certainly not sure, I suspect that there still is need for something like Postgres.

Also: I will most likely be able to contribute to this effort if we move forward.

**Comment by furuholm** [comment]
To get things going I have created an experimental fork that exports all vulnerabilities from Clair to notes in Grafeas. I chose to export rather than using Grafeas as a backend because of the reasons I stated above. Any thoughts on that decision?

https://github.com/combient/clair/tree/experimental_grafeas_export

To test it out: change the configuration (see sample config file) and start Clair. The export process is executed after the vulnerability update has finished.

I have probably missed some corner cases and things are probably not in the right place, so please consider this as input to the process rather than a production ready PR. (+ there is only support for the Postgres backend and I have not written tests)

**Comment by R2wenD2** [comment]
@furuholm @jzelinskie what are your thoughts on creating a new storage implementation that uses Grafeas for this? 

We have an issue in Grafeas to have an audit trail (https://github.com/Grafeas/Grafeas/issues/2) which would allow us to insert the notification message in Grafeas. 

We'll probably want an actual design doc - and perhaps a joint datastore that allows one to use both existing clair and Grafeas datastores while the Grafeas api is built out to support audit tracking.

**Comment by R2wenD2** [comment]
@clrprod might have some feedback here as well.

**Comment by clrprod** [comment]
Having clair write directly to grafeas for vulnerabilities would be pretty awesome!

I agree with furuholm@ that notifications likely wouldn't work (there is some work, but I don't think it's live in grafeas yet). I also think a whole other bunch of stuff would break (i.e. it looks like there is a layer storage system and ancestry which may be moveable to grafeas), but we likely don't want to move the arbitrary key value store to grafeas.

I know that we use clair by actually creating our own binary that implements the internal structures on grafeas (although we use clair v0.13.6 so it's unlikely to be directly portable). We directly pass a DataStore to the clair Updater where only the vulnerability writing functionality is implemented.

If this is a pretty common use case, maybe we could make a updater only version of the clair binary that writes to grafeas, or have flags / config that can make the normal clair binary not do all of the other functionality.

**Comment by jzelinskie** [comment]
@R2wenD2 I see a few strawman integrations:

## 1. Clair implements its storage interface by calling the Grafeas API.

If it is even possible to store all the metadata Clair in Grafeas, I worry that replacing Postgres with Grafeas directly for storage would cause a significant performance degradation for the Clair gRPC API. There's a [fairly relational object model](https://github.com/coreos/clair/blob/1425086670521ed5510e5dadd5e0b374d33e0b87/database/database.go#L81) used for storage in Clair that if implemented with Grafeas, I don't believe could respond to user's requests with interactive latency. This model understands container filesystem layers and abuses content addressability to reduce repeated computation. We also _just_ designed this model for Clair v3 and would like to get it in the hands of our users before considering another complete rewrite.

## 2. Clair implements the Grafeas API using its own storage.

Do you have a storage layer besides the sample memory store? It seems like it makes more sense for Clair to implement the Grafeas API using its own storage engine and expose both APIs. This leaves Clair in a scenario where we could only expose a read-only Grafeas API or allow change the scope of the Clair project to include managing all the metadata Grafeas does.

## 3. Clair implements a worker to synchronize with a Grafeas API/database

This is what @furuholm has been implementing as a strawman. I think this gets us a lot of bang for our buck, but is probably not the best long-term solution because of the many moving parts.

**Comment by furuholm** [comment]
@jzelinskie: some thoughts

1: Agree.

2: I hadn't thought about this option. Not a bad idea! I worry that the Grafeas API would step on Clairs' toes though. What would it mean to add vulnerability notes from the Grafeas API for example? Would those go straight into the Vulnerability table? How would that affect data integrity (like namespace mapping)? The vulnerability notes could go into its own table, but then the whole premise of the idea is kind of lost. 

But maybe a subset would work? I'm thinking exposing a read-only version of vulnerability notes would allow CVE:s to be imported into a Grafeas server by a third party or by the Grafeas server itself. Alternatively the Grafeas server could proxy vulnerability requests to Clair. I haven't thought the occurrences through, but my guess is that a read only API could work for this as well. (If you don't want to make Clair Grafeas compliant it could be enough to expose the data needed through some other API).

3: Agree

Either way I would love for there to be an open source grafeas server that was built using the same architecture/structure as Clair, but I'm not sure they should be the same.

**Comment by R2wenD2** [comment]
Agree that 3 is a good short term model, and there's no real reason to not do that now since we all agree on this approach as a short term solution.

Option 2 is interesting. I think there might be room for a hybrid of 2. We don't need to expand the scope of clair to handle all metadata types supported by Grafeas in order for it to support a Read/Write Grafeas. Consider a "bring your own Grafeas" solution where when running Clair, you can point it at an existing Grafeas instance and have it create vulnerability notes/occurrences there.

@jzelinskie and @furuholm are you both around next week? Would you be open for a quick meeting? It might be easier for us to brainstorm over a vc.

**Comment by furuholm** [comment]
@R2wenD2 Vc sounds good to me and I'm around all of next week. You have my mail address.

**Comment by jzelinskie** [comment]
I'm also open to VC next week.

**Comment by furuholm** [comment]
Some more input to the discussion/the planned vc. This time about occurrences.

It seems reasonable to consider a one-to-one mapping between vulnerabilities in Clair and Grafeas notes. The project to which notes shall be written can be a single config parameter as this project will be shared by all users. Detected vulnerabilites in containers however are user or customer specific. So to what project should a vulnerability occurrence be reported? The mapping can be done for Google's container scanner since Google Cloud project ids are used for both the container registry and the container scanning. But I'm not sure this mapping is possible to do in the general case.

**Comment by jzelinskie** [comment]
@furuholm I've made a branch [experiment-grafeas-sync](https://github.com/coreos/clair/tree/experiment-grafeas-sync) which we can use as a target for a PR of your PoC to for safe keeping. This way we can get some code review through the normal processes.

**Comment by furuholm** [comment]
@jzelinskie PR created.

**Comment by furuholm** [comment]
@jzelinskie during the VC in december we discussed developing a proposal based on a pubsub architecture. Do you have any suggestions or preferences regarding pubsub technologies? 
/cc @R2wenD2

**Comment by furuholm** [comment]
Come to think of it I think we need a discussion about the expectations before we can do much about this. 

Some questions that I think needs to be addressed:
- Are there consumer specific queues like the notification system in Clair today or should queues be shared? Are late joiners supported? 
- What happens when you blow the database (that is basically what is happening during migration as far as i understand). Are all the CVE:s reposted?
- Should a specific technology be targeted or is a pluggable system preferred?
There is probably more stuff that needs to be addressed.

Also: this issue is probably not the right medium for this discussion. Should I (or someone else) open a new issue?

**Comment by R2wenD2** [comment]
@JustinBeckwith do you have any thoughts on this at all?

**Comment by JustinBeckwith** [comment]
Greetings folks!  Still very new to the project, and taking things in - but in general here are my thoughts:
- Integration between Grafeas and Clair is a big win for everyone, and I'm super excited about this 😁
- I also worry about trying to use Grafeas as a general purpose data storage layer.  I wouldn't expect most folks to move from Postgres to Grafeas, rather publish metadata to Grafeas as part of an existing workflow.
- We should have a follow up discussion with @R2wenD2 and myself about the the questions above.
- Feel free to open a new issue specifically for this discussion!

**Comment by R2wenD2** [comment]
@furuholm @jzelinskie Would it be easier to schedule another VC to try to address the discussion points @furuholm mentions above?

**Comment by furuholm** [comment]
I've tried to outline a proposal. Thoughts?

**Problem statement:**
Allow third party systems (like Grafeas) to continuously ingest Clair CVE listings and ancestry data (i.e. container image metadata and their identified vulnerabilities).

**Requirements:**
- Allow access to current state as well as state changes such as added/removed/changed CVE:s and  ancestries.
- Support multiple delivery mechanisms.

**Proposed solution:**
- Clair exposes a (golang) extension point for notifications that is triggered when changes occur. The extension is implemented and configured like the existing Clair extensions. 
Potential implementations: 
  * message busses like amqp, 
  * pub/sub systems like Kafka or Google Cloud Pub/Sub, 
  * the current Clair notification API?
  * ...
- Clair exposes an API that enables users to communicate with the notification extensions to allow e.g. create/delete notification stream. This would 
  * provide a way for external systems to be late joiners. 
  * allow multiple independent consumers to coexist 

Not having an API for creating streams would probably require separate mechanisms for accessing the current state and updates as well as some way to synch these two to ensure data consistency.

**Open questions:**
- How should db migrations be handled? From reading the code it looks like clair creates new tables and drops old ones. This will then be followed by populating the db with new data that may or may not match the state of the previous migration. I'm thinking requiring the consumers to be idempotent (accepting that the same CVE or vulnerability to be delivered more than once) could make this a non-issue. Thoughts on this?

**Example use case:**
- Client `A` uses the API to create a notification stream `S`. 
- The notification extension instance starts by writing/publishing the current state to `S`
- The notification extension instance waits for changes and writes/publishes those to `S` as they occur
- `C` consumes the data through `S`

**Comment by thomasmckay** [comment]
If I understand the clair's purpose correctly, multiple sets of input data comes in through drivers. After that, clair has the tooling to inspect an image and cross-reference it with the input data. Why does it have any part in notifying additional tools? (Not being snarky here, just trying to understand all the pieces.:) )

For grafeas I could see a general driver being created that took grafeas api as input. In turn, I could also see having a grafeas api endpoint for querying.

As an example, say there is a tool that produces errata-to-package information (like the existing drivers today). It's api is grafeas and can thus readily be an input driver for clair. Next an image is scanned by clair and associations made to this data. Let's say the tooling that initiated the scan prefers results back in grafeas format. Perfect.

Now later, the errata-to-package info is updated in clair via the periodic driver. Is the proposal that clair should output information to a stream? What would the stream output? If the consumer of the stream is getting the errata-to-package updates, shouldn't they go get the results directly from source instead? If the consumer is expecting to get updated scan results (ie. a new errata shows up for a previously scanned image) how would that scale?

**Comment by furuholm** [comment]
This is not about making grafeas an input source to Clair. It is about making Grafeas a consumer of the data that Clair outputs. The consumer in the use case I laid out above could in the Grafeas case be either Grafeas itself or a tool that writes the reported changes to Grafeas.

Does this make sense or are we talking past each other?

**Comment by thomasmckay** [comment]
I think I understand. In the experimental grafeas branch, for example, as vulnerabilities are imported they are in turn exported to a grafeas server (via grafeas protobuf api). Your proposal is working towards defining how that would work.

I see that my suggestion of a "grafeas input driver" is a separate idea, though perhaps could benefit from the register notion for late joiners.

**Comment by staylorx** [comment]
I use Clair as a CI gate in my deployment chain and I like it quite a lot. I'm starting to pull Grafeas into my work. I think the idea of populating notes and occurrences with Clair-generated information is interesting. 

I read in some of the comments above the idea of replacing the postgres backend datastore with Grafeas but I'm not sure it works that way. The reference implementation doesn't have a persistent backend and the container-based implementation I've looked at (Anchore-engine) use a database. 

It seems that this is a matter of having a server in front of Clair proper with the Grafeas API implemented in addition to the existing Clair services? A risk to using the Grafeas API as the backend driver of Clair is that "vulnerabilty" is only one of six of the Grafeas "kinds" of note. But as an optional flavour of Clair API I think it'd be a great add.

**Comment by R2wenD2** [comment]
We abandoned the concept of Grafeas backing Clair fairly quickly. @furuholm has a [fork](https://github.com/combient/clair/tree/experimental_grafeas_export) which feeds data from Clair into Grafaes: https://github.com/combient/clair/tree/experimental_grafeas_export that might do what you want. 

The Grafeas reference implementation does have a postgres datastore now as well.

**Comment by staylorx** [comment]
Thanks for the update. Good stuff. I'll check that fork.

**Comment by thomasmckay** [comment]
Maybe I'm wrong in this thinking but I view grafeas as an API definition. As long as the backend service can massage the data to fit the API format, things are good to go. Rather than transfer data to an external grafeas database, could an API be put onto clair to support the portions of grafeas that match its data?

@staylorx I like the idea of clair being part of the toolchain and having an output (and probably input) source be grafeas compatible.

**Comment by bansalrajat** [comment]
@furuholm  I just tried to integrate the fork with the grafeas instance i have and i am not getting anything. Any document on steps need to be followed.

**Comment by furuholm** [comment]
@bansalrajat unfortunately I would say that the published fork is a dead end. After discussing this further we (me, @R2wenD2 and @jzelinskie) agreed that the way to solve this problem is probably to use a "clair proxy" that forwards scanning requests to clair and stores the result in grafeas as results are returned to the user. This is similar to how Quay works (or at least how it worked at the time of these discussions). Unfortunately the company I work for changed focus about the time we had concluded this, so I never got around to implement it ☹️

**Comment by bansalrajat** [comment]
@furuholm I was trying to integrate grafeas with harbor so the vulnerabilities data could be loaded into grafeas. But seems like it wouldn't be possible for some time. Thanks for the update BTW. Cheers.

**Comment by hdonnay** [comment]
We’re declaring bug bankruptcy as part of the release process for a new major version of Clair. Please open a ticket in our [issue tracker](https://issues.redhat.com/projects/PROJQUAY/) if you feel this still needs to be addressed, and we'll triage as part of our v4 development process. Thanks!
